vllm/compilation/backends.py

```diff
@@ -164,80 +166,212 @@ def wrap_inductor(graph, example_inputs, additional_inductor_config):
+def vllm_backend(graph, example_inputs) -> Callable:
+
+    # config is read when we first compile the graph
+    # (i.e. when this backend is first called)
+    compilation_configs = CompilationConfig.default_config()
+
+    from vllm.plugins import get_non_cudagraph_ops
+    non_cudagraph_ops = get_non_cudagraph_ops()
+
+    from torch._dynamo.utils import lazy_format_graph_code
+
+    # split graph by non_cudagraph_ops
+    subgraph_id = 0
+    node_to_subgraph_id = {}
+    attention_graphs = []
+    for node in graph.graph.nodes:
+        if node.op in ("output", "placeholder"):
+            continue
+        if node.op == 'call_function' and str(
+                node.target) in non_cudagraph_ops:
+            subgraph_id += 1
+            node_to_subgraph_id[node] = subgraph_id
+            attention_graphs.append(subgraph_id)
+            subgraph_id += 1
+        else:
+            node_to_subgraph_id[node] = subgraph_id
+
+    graph_pool = torch.cuda.graph_pool_handle()
+    final_graph: Callable = None  # type: ignore
+
+    if subgraph_id != 0:
+        # `keep_original_order` is important!
+        # otherwise pytorch might reorder the nodes and
+        # the semantics of the graph will change when we
+        # have mutations in the graph
+        split_gm = torch.fx.passes.split_module.split_module(
+            graph,
+            None,
+            lambda node: node_to_subgraph_id[node],
+            keep_original_order=True)
+
+        logger.debug("%s", lazy_format_graph_code("stiching module", split_gm))
+
+        # sort the names to make sure the order is deterministic
+        names = [name for (name, module) in split_gm.named_modules()]
+        names.sort()
+
+        is_first_graph = True
+        for name in names:
+            if "." in name or name == "":
+                # recursive child module or the root module
+                continue
+
+            module = getattr(split_gm, name)
+
+            graph_id = int(name.replace("submod_", ""))
+            if graph_id not in attention_graphs:
+                # cannot setattr to a module, so we need to set it to the dict
+                split_gm.__dict__[name] = piecewise_backend(
+                    module, compilation_configs, graph_pool, is_first_graph)
+                is_first_graph = False
+
+        final_graph = split_gm
+
+    else:
+        final_graph = piecewise_backend(graph,
+                                        compilation_configs,
+                                        graph_pool,
+                                        is_first_graph=True)
+
+    # trigger the first compilation
+    final_graph(*example_inputs)
+
+    return final_graph
+
+
+@dataclasses.dataclass
+class Entry:
+    runnable: Callable
+    use_cudagraph: bool
+    target_warmup_times: int
+    current_warmup_times: int = 0
+    cudagraph: Optional[torch.cuda.CUDAGraph] = None
+    output: Optional[Any] = None
+
+
+def piecewise_backend(graph,
```

Comment:
A short docstring that describes the piecewise_backend function would be helpful. A lot of stuff are going on inside.

Reply:
will address together with @ProExpertProg 's comment below.